backbone.js

```diff
@@ -1079,10 +1079,26 @@
       if (!wait) this.add(model, options);
       var collection = this;
       var success = options.success;
+      var forwardPristineError;
       options.success = function(m, resp, callbackOpts) {
-        if (wait) collection.add(m, callbackOpts);
+        if (wait) {
+          model.off('error', forwardPristineError);
+          collection.add(m, callbackOpts);
+        }
         if (success) success.call(callbackOpts.context, m, resp, callbackOpts);
       };
+      // In case of wait:true, our collection is not listening to any
+      // of the model's events yet, so it will not forward the error
+      // event. In this special case, we need to listen for it
+      // separately and handle the event just once.
+      // (The reason we don't need to do this for the sync event is
+      // in the success handler above: we add the model first, which
+      // causes the collection to listen, and then invoke the callback
+      // that triggers the event.)
+      if (wait) {
+        forwardPristineError = _.bind(this._onModelEvent, this, 'error');
```

Comment:
this could be..  (forgive the ES6)
```js
forwardPristineError = (...args) => {
  if (this.has(model) return;
  this._onModelEvent('error', ...args);
}
```
I _think_ this prevents the double error.. it handles when the model doesn't yet have the duplicate id.. _and_ it moves the extra check into the corner-case handler... But I could be missing something?

Reply:
well if the model is considered new at add it'd be a different `cid` and then the server can potentially return an existing model id.. so pre-wait `has` would be false, but post `wait` it would be `true`

Reply:
Wait. (No punt intended.) The `cid` is completely independent from newness and server data, no?

Also, if the model is not in the collection beforehand and the request errors, its attributes shouldn't change.

Reply:
Ha well you're right on the error part.. since there's an error the attributes won't change, so there's no real issue.. _but_ the `has` checks both `model.cid` and `model.id` so it's possible to have two models with different `cid`s but the same `id` and it won't add the 2nd.. but seeing as the `id` can't change before the `error` event is thrown, it's a mute point.  So I guess the only benefit is delaying the has check until the last necessary moment?  I guess that's worth it...